Wrappers/Python/cil/optimisation/functions/TotalVariation.py

```diff
@@ -188,36 +253,46 @@ def projection_P(self, x, out=None):
     
     def proximal(self, x, tau, out = None):
         
-        ''' Returns the solution of the FGP_TV algorithm '''         
+        r""" Returns the proximal operator of the TotalVariation function at :code:`x` ."""        
         try:
             self._domain = x.geometry
         except:
             self._domain = x
+
+        # Compute Lipschitz constant provided that domain is not None.
```

Comment:
with the above try/except can domain ever be none?

Reply:
At the instance level, `_domain` is `None`.  In order to use the `call` and `proximal` methods, we need to define the `GradientOperator` hence its geometry. The try/except is due to SIRF/CIL integration. So in the SIRF case `x.geometry` does not exist and will get a `SIRF DataContainer`.

We discussed this with @paskino and decide  to `raise ValueError` in the `gradient` property

https://github.com/TomographicImaging/CIL/blob/3f437ae028fa6ca37270ede688ea6f56726e7bbe/Wrappers/Python/cil/optimisation/functions/TotalVariation.py#L368-L373

So if we create a TotalVariation instance and `call` or `proximal` methods are not used before this will raise an error.


```python

    tv = TotalVariation()
    res = tv.gradient

```

```bash
ValueError:  The domain of the TotalVariation is None. Please use the __call__ or proximal methods first before calling gradient.
```